### PR TITLE
Expose 'with default RSpec/Language config' to consumers

### LINKED ISSUE #1418: Idea: make spec support available in the gem
Hello,

I was trying to extend rubocop-rspec with a few additional rules to enforce coding practices we have in my team, and I ran into the fact that I cannot include the `spec_helper` and shared contexts from the gem to test my extension.

What rubocop does is that they have their rspec support tools (contexts, expect_offenses and so on) defined in a subdirectory of their [`lib` folder ](https://github.com/rubocop/rubocop/tree/master/lib/rubocop/rspec) so that this tooling is available when you `require 'rubocop'` in your Gemfile.

Would you agree to a similar approach in `rubocop-rspec`? Happy to make a PR if so.

### PR DESCRIPTION
**Pull request by smcgivern**
This helper is essential for anyone writing their own RSpec cops based on `RuboCop::Cop::RSpec::Base`, because it now requires a minimal config to even run the cop.

The wordy file name is because the RuboCop gem already defines `rubocop/rspec/shared_contexts`, and we don't want to shadow that. This means that consumers of this library who want to write a spec for a cop based on `RuboCop::Cop::RSpec::Base` will need something like this in `spec/spec_helper.rb`:

```ruby
require 'rubocop/rspec/shared_contexts/default_rspec_language_config_context'

RSpec.config do |config|
  # ...
  # Add metadata as appropriate
  config.include_context 'with default RSpec/Language config'
end
```

This was suggested in the PR that introduced this feature: https://github.com/rubocop/rubocop-rspec/pull/956/files#r515643915

> We don't enforce using alias configuration, and custom cops can continue to use their own configuration until someone decides to follow the mainstream, and would like to contribute by extracting this somewhere to lib/.

fixes #1418
---

Before submitting the PR make sure the following are checked:

* [x] Feature branch is up-to-date with `master` (if not - rebase it).
* [x] Squashed related commits together.
* [ ] Added tests.
* [x] Updated documentation.
* [x] Added an entry to the `CHANGELOG.md` if the new code introduces user-observable changes.
* [x] The build (`bundle exec rake`) passes (be sure to run this locally, since it may produce updated documentation that you will need to commit).

### CODE REVIEW

**Comment by smcgivern** [comment]
>  Updated documentation.

I'd like to do this, but I wasn't sure where. Maybe https://github.com/rubocop/rubocop-rspec/blob/master/docs/modules/ROOT/pages/upgrade_to_version_2.adoc?

**Comment by smcgivern** [comment]
I tested this on a library we have that uses rubocop-rspec and it worked for me: https://gitlab.com/gitlab-org/ruby/gems/gitlab-styles/-/merge_requests/119#note_1092839680

File: spec/spec_helper.rb [hunk 1]
```diff
@@ -7,7 +7,7 @@ module SpecHelper
   ROOT = Pathname.new(__dir__).parent.freeze
 end
 
-spec_helper_glob = File.expand_path('{support,shared}/*.rb', __dir__)
+spec_helper_glob = File.expand_path('{support,shared,../lib/rubocop/rspec/shared_contexts}/*.rb', __dir__)
```
Comment (smcgivern) [inline spec/spec_helper.rb:10]:
```
spec/spec_helper.rb:10:81: C: [Correctable] Layout/LineLength: Line is too long. [106/80]
spec_helper_glob = File.expand_path('{support,shared,../lib/rubocop/rspec/shared_contexts}/*.rb', __dir__)                                                                                ^^^^^^^^^^^^^^^^^^^^^^^^^^
```

Whoops, I guess I'll break this into multiple lines.

**Comment by pirj** [comment]
> Maybe https://github.com/rubocop/rubocop-rspec/blob/master/docs/modules/ROOT/pages/upgrade_to_version_2.adoc?

Yes, or add a section similar to RuboCop's https://docs.rubocop.org/rubocop/development.html

**Review by pirj — APPROVED** [review]
Thank you!
Appreciate if you could add some docs.

**Comment by smcgivern** [comment]
Thanks @pirj, I added some brief development documentation. (Rendered: https://github.com/smcgivern/rubocop-rspec/blob/ad674c8ee2a55a48f933a4605d9145daaa83c7ba/docs/modules/ROOT/pages/development.adoc)

I am not sure what the documentation guidelines are here, so I'm happy to take any and all suggestions there.

**Comment by bquorning** [comment]
I have renamed some of the required CI checks, which unfortunately means that you will need to rebase this branch to make CI pass.

**Comment by smcgivern** [comment]
Thanks, done.

File: docs/modules/ROOT/pages/development.adoc [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+
+[source,ruby]
+----
+require 'rubocop/rspec/shared_contexts/default_rspec_language_config_context'
+
+RSpec.config do |config|
+  # ...
+  # Add metadata as appropriate
+  config.include_context 'with default RSpec/Language config'
+end
```
Comment (pirj) [inline docs/modules/ROOT/pages/development.adoc:23]:
```ruby
  # Set metadata on all cop specs
  config.define_derived_metadata(file_path: %r{/spec/rubocop/cop/}) do |meta|
    meta[:type] = :cop_spec
  end

  # Include config shared context for all cop specs
  config.define_derived_metadata(type: :cop_spec) do |meta|
    meta[:config] = true
  end

  config.include_context 'with default RSpec/Language config', :config
```

Notice the difference that the context that defines `config` will only be included in specs from certain directories, or explicitly marked as `:cop_spec`.

This is only important for monolithic projects that also define cops among all, not to litter **all** specs with the populated `config`.
For custom RuboCop extensions that inspect specs, your current approach would work perfectly fine.

**Comment by tdeo** [comment]
Fixes #1418

File: docs/modules/ROOT/pages/development.adoc (hunk 2 above)
Comment (smcgivern) [reply docs/modules/ROOT/pages/development.adoc:31]: @pirj this is what I was getting at with 'add metadata as appropriate', but I agree it's better to be clear here. I've added this change and a note that it's an example above the code block.

**Comment by smcgivern** [comment]
@tdeo thanks for trying it out!

I rebased because the CHANGELOG had conflicts, but the workflows need a maintainer to approve before they run:

image

File: CHANGELOG.md [hunk 3]
```diff
@@ -3,6 +3,7 @@
 ## Master (Unreleased)
 
 - Add `named_only` style to `RSpec/NamedSubject`. ([@kuahyeow])
+* Add `with default RSpec/Language config` helper to `lib` (under `rubocop/rspec/shared_contexts/default_rspec_language_config_context`), to allow use for downstream cops based on `RuboCop::Cop::RSpec::Base`. ([@smcgivern])
```
Comment (pirj) [inline CHANGELOG.md:6]:
```suggestion
- Add `with default RSpec/Language config` helper to `lib` (under `rubocop/rspec/shared_contexts/default_rspec_language_config_context`), to allow use for downstream cops based on `RuboCop::Cop::RSpec::Base`. ([@smcgivern])
```

File: CHANGELOG.md [hunk 4]
```diff
@@ -31,6 +32,9 @@
 - Do not attempt to auto-correct example groups with `include_examples` in `RSpec/LetBeforeExamples`. ([@pirj])
 - Add new `RSpec/SortMetadata` cop. ([@leoarnold])
 - Add support for subject! method to `RSpec/SubjectDeclaration`. ([@ydah])
+* Add `require_implicit` style to `RSpec/ImplicitSubject`. ([@r7kamura][])
+* Fix a false positive for `RSpec/Capybara/SpecificMatcher` when `have_css("a")` without attribute. ([@ydah][])
+* Add `with default RSpec/Language config` helper to `lib` (under `rubocop/rspec/shared_contexts/default_rspec_language_config_context`), to allow use for downstream cops based on `RuboCop::Cop::RSpec::Base`. ([@smcgivern][])
```
Comment (pirj) [inline CHANGELOG.md:37]:
```suggestion
```

**Comment by pirj** [comment]
Green!

**Review by Darhazer — APPROVED** [review]

**Comment by pirj** [comment]
@smcgivern May I ask you to rebase and make sure Changelog is in order?

**Comment by smcgivern** [comment]
@pirj sorry for the push spam, I didn't fix the CHANGELOG properly first time around. This is up to date now.

**Comment by pirj** [comment]
Thank you, @smcgivern !

**Comment by smcgivern** [comment]
Thanks @pirj! I look forward to deleting this file from our repos when this is release 😃

**Comment by pirj** [comment]
Released in [2.15.0](https://github.com/rubocop/rubocop-rspec/releases/tag/v2.15.0), @smcgivern

**Pull request merged**